cmd/cli/bulk.go

```diff
@@ -0,0 +1,185 @@
+package main
+
+import (
+	"bufio"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"log"
+	"os"
+	"strings"
+
+	"github.com/multiplay/go-svrquery/lib/svrquery"
+	"github.com/multiplay/go-svrquery/lib/svrquery/protocol"
+)
+
+const (
+	numWorkers = 100
+)
+
+var (
+	errNoItem       = errors.New("no item")
+	errEntryInvalid = errors.New("invalid entry")
+)
+
+// BulkResponseItem contains the information about the query being performed
+// against a single server.
+type BulkResponseItem struct {
+	Address    string                      `json:"address"`
+	ServerInfo *BulkResponseServerInfoItem `json:"serverInfo,omitempty"`
+	Error      string                      `json:"error,omitempty"`
+}
+
+// BulkResponseServerInfoItem container the server information.
+type BulkResponseServerInfoItem struct {
+	CurrentPlayers int64  `json:"currentPlayers"`
+	MaxPlayers     int64  `json:"maxPlayers"`
+	Map            string `json:"map"`
+}
+
+// BulkResponseItemWork is an item returned by an worker containing the data item
+// plus any terminal error it encountered.
+type BulkResponseItemWork struct {
+	Item *BulkResponseItem
+	Err  error
+}
+
+// queryBulk queries a bulk set of servers using a query file.
+func queryBulk(file string) error {
+	f, err := os.Open(file)
+	if err != nil {
+		log.Fatal(err)
+	}
+	defer f.Close()
+
+	// Make a jobs channel and a number of workers to processes
+	// work off of the channel.
+	jobChan := make(chan string)
+	resultsChan := make(chan BulkResponseItemWork)
+	for w := 1; w <= numWorkers; w++ {
+		go worker(jobChan, resultsChan)
+	}
+
+	items := make([]BulkResponseItem, 0, 1000)
+
+	// Queue work onto the channel
+	scanner := bufio.NewScanner(f)
+	jobCount := 0
+	for scanner.Scan() {
+		jobCount++
+		jobChan <- scanner.Text()
+	}
+
+	// Receive results from workers.
+	for i := 0; i < jobCount; i++ {
+		v := <-resultsChan
+		switch {
+		case errors.Is(v.Err, errNoItem):
+			// Not fatal, but no response for this entry was created.
+			continue
+		case v.Err != nil:
+			// We had a major issue, force immediate stop.
+			return fmt.Errorf("fatal error from worker: %w", v.Err)
+		}
+
+		// add the item to our list of items.
+		items = append(items, *v.Item)
+	}
+
+	b, err := json.MarshalIndent(items, "", "\t")
+	if err != nil {
+		return err
+	}
+	fmt.Printf("%s\n", b)
```

Comment:
Debug line

Reply:
In this case its not debugging as its the main print to stdout in this bulk mode